### PR TITLE
 feat: move to a single-export package pattern for libs

### PR DESCRIPTION
**Pull request by varl**
Several excellent points have been brought up related to the complexity
of how our libraries are built/deployed/published. A move towards a
single-export package structure simplifies several aspects of our
libraries:

It forces developers to use the entry-point in the `package.json` file,
and to use named imports for the components.

- This allows Jest to use CJS for tests _(verified with usage-analytics-app)_

- This allows Webpack to use strict ESM

- This allows us to publish from the root package and deprecate the
  `publish-build` script

- Tree-shaking will work since Webpack resolves to the `module` field
  and will use the ESM build _(verified in CRA2.0)_

- It will be easier to search through our polyrepo codebase for usages
  as the named export must be referenced as is defined

BREAKING CHANGE: All imports must move to named imports from the
entry-point.

### CODE REVIEW

File: docs/src/App.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ import Content from './Content'
 
 import 'typeface-roboto/index.css'
 
-import 'css/reset.css'
+import '@dhis2/ui-core/build/es/css/reset.css'
```
Comment (varl) [inline docs/src/App.js]: This is not great. Gonna solve this in a different way.

**Comment by varl** [comment]
This migrates the `@dhis2/ui` library to `@dhis2/ui-core`, cuts out the `api.js` class, cuts out the HeaderBar (it's due for `@dhis2/ui-widgets` anyway) and flattens the entire package into a single-entry point package.

It does not provide a default export so everything needs to be accessed by named imports.

cc: @ismay @amcgee @HendrikThePendric

**Comment by varl** [comment]
The built package would look like:

```
ui-core/
├── build
│   ├── cjs
│   │   ├── Button
│   │   ├── Card
│   │   ├── Checkbox
│   │   ├── Chip
│   │   ├── CircularProgress
│   │   ├── css
│   │   ├── Divider
│   │   ├── DropdownButton
│   │   ├── Help
│   │   ├── icons
│   │   ├── index.js
│   │   ├── InputField
│   │   ├── LinearProgress
│   │   ├── Logo
│   │   ├── Menu
│   │   ├── Radio
│   │   ├── SelectField
│   │   ├── SplitButton
│   │   ├── Switch
│   │   ├── theme.js
│   │   └── utils
│   └── es
│       ├── Button
│       ├── Card
│       ├── Checkbox
│       ├── Chip
│       ├── CircularProgress
│       ├── css
│       ├── Divider
│       ├── DropdownButton
│       ├── Help
│       ├── icons
│       ├── index.js
│       ├── InputField
│       ├── LinearProgress
│       ├── Logo
│       ├── Menu
│       ├── Radio
│       ├── SelectField
│       ├── SplitButton
│       ├── Switch
│       ├── theme.js
│       └── utils
├── CHANGELOG.md
├── LICENSE
├── package.json
└── README.md
```

NOTE: **This disallows _reaching into the internals of a package and importing a specific file_.**

This is an important side-effect and a limitation which forces us to put our modules in the correct package. If it doesn't make sense to have it in the entry-point, it doesn't make sense that it should be exposed from the package, and it needs a new home.

E.g. the `utils/` module. Since it is possible to reach in and grab it, that has become a pattern we do not necessarily want to promote, as it makes it harder for us to change the internals of a package and maintain the external API.

**Comment by amcgee** [comment]
Looks like the CSS export is broken, as mentioned above.  From netlify build:

```
Module not found: Error: Can't resolve '@dhis2/ui-core/build/es/css/reset.css' in '/opt/build/repo/docs/src'
```

Perhaps non-js exports should exist outside the src directory and therefore be accessible by file reference?  Something like `import '@dhis2/ui/css/reset.css'`.  They then become effectively part of the exposed API, so we'd be committing to maintining those files.  Another option could be to `export { default as cssReset } from 'reset.css'`, but I'm not sure that does the right thing.  Might require a js wrapper file which just imports the `reset.css` file?

**Comment by amcgee** [comment]
There's an unrelated, cryptic-looking error in Travis as well.

**Comment by varl** [comment]
> Perhaps non-js exports should exist outside the src directory and therefore be accessible by file reference? Something like `import '@dhis2/ui/css/reset.css'`. They then become effectively part of the exposed API, so we'd be committing to maintining those files. Another option could be to `export { default as cssReset } from 'reset.css'`, but I'm not sure that does the right thing. Might require a js wrapper file which just imports the `reset.css` file?

I'm pondering moving it out from the `src` folder and adding it to the `files` array to put it in the bundle as: `@dhis2/ui-core/css/reset.css`.

But.

Buuuut.

We don't need to bundle the `reset.css` file at all and leave it to the Apps, it's not required from the components side anymore (after the port to `styled-jsx`)

File: docs/src/App.js (hunk 1 above)
Comment (amcgee) [reply docs/src/App.js]: See my PR comment for another approach to this

**Comment by varl** [comment]
> There's an unrelated, cryptic error in Travis as well.

The docs used to just reach into the build directory, but I used a `link` now to make sure the imports work as expected to the root package. One sec.

File: package.json [hunk 2]
```diff
@@ -50,5 +50,8 @@
   "dependencies": {
     "classnames": "^2.2.6",
     "styled-jsx": "^3.2.1"
-  }
+  },
+  "files": [
+    "./build"
```
Comment (amcgee) [inline package.json]: If we expose non-js files directly in, i.e., a `./css` folder we could add that folder to the `files` array and, if we want to get fancy, enforce that removing a file from a non-build published folder is automatically a breaking change.

**Comment by varl** [comment]
> Something like `import '@dhis2/ui-core/css/reset.css'`. They then become effectively part of the exposed API, so we'd be committing to maintining those files. Another option could be to `export { default as cssReset } from 'reset.css'`

I think this requires use to have css-loader and file-loader set up in Webpack when the lib is pulled in. One of the reasons we replaced PostCSS with styled-jsx was to get away from that.

**Comment by amcgee** [comment]
> I think this requires use to have css-loader and file-loader set up in Webpack when the lib is pulled in. One of the reasons we replaced PostCSS with styled-jsx was to get away from that.

Even better - export a `<ResetCSS />` component which is just a styled-jss port of `reset.css` using `<style global>{...reset.css...}</style>`

**Comment by varl** [comment]
> 
> 
> > I think this requires use to have css-loader and file-loader set up in Webpack when the lib is pulled in. One of the reasons we replaced PostCSS with styled-jsx was to get away from that.
> 
> Even better (if we want apps to be able to use a reset, which could be helpful for things like body margins and border-box and things) - export a `<ResetCSS />` component which is just a styled-jsx port of `reset.css` using `<style global>{...reset.css...}</style>`
> 
> = SIDE-EFFECT FREE!

Adding it now. 💯

File: src/CssReset/index.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
+import React from 'react'
+
+function CssReset() {
+    return (
+        <style jsx global>{`
+            /* http://meyerweb.com/eric/tools/css/reset/
```
Comment (Mohammer5) [inline src/CssReset/index.js]: I think we can move that outside of the style tag, no need for styled-jsx do strip that?

Comment (varl) [reply src/CssReset/index.js]: Agreed.

**Comment by varl** [comment]
This also introduces `semantic-release` which is a fully automated release process based on conventional commits, so each merge to master results in a release.

This puts further pressure on branches being well-tested.

**Comment by ismay** [comment]
> It forces developers to use the entry-point in the package.json file, and to use named imports for the components. 

> A move towards a single-export package structure simplifies several aspects of our libraries

> This disallows reaching into the internals of a package and importing a specific file.

> flattens the entire package into a single-entry point package.

Completely agree. I prefer a libs file structure to be private, this solution seems perfect to me. The deep imports we do with libs like d2 for example are something I'd really like to move away from, so I am super happy with this change 🕺

> This is an important side-effect and a limitation which forces us to put our modules in the correct package. If it doesn't make sense to have it in the entry-point, it doesn't make sense that it should be exposed from the package, and it needs a new home.

> E.g. the utils/ module. Since it is possible to reach in and grab it, that has become a pattern we do not necessarily want to promote, as it makes it harder for us to change the internals of a package and maintain the external API.

Exactly, 100% agree. One thing I was wondering though, what do you mean here with: it needs a new home? Since it's relied on by people, I assume you mean that it should be published as a separate package?

**Comment by varl** [comment]
> One thing I was wondering though, what do you mean here with: it needs a new home? Since it's relied on by people, I assume you mean that it should be published as a separate package?

I meant that if a module shouldn't be exposed from the library entry-point, but we want to re-use it, then it needs to be placed somewhere else, might be another package we have or a new package if nothing else makes sense.

**Comment by ismay** [comment]
> I meant that if a module shouldn't be exposed from the library entry-point, but we want to re-use it, then it needs to be placed somewhere else, might be another package we have or a new package if nothing else makes sense.

Ok, yeah exactly, I agree.

**Comment by varl** [comment]
> 
> 
> One thing I'd like to add, I'd recommend going with normalize.css over a reset.
> 
> > The Normalize.css stylesheet also removes browser inconsistencies for HTML elements, but instead of removing everything like CSS Reset, normalize.css will preserve some useful defaults. Normalize.css was created to make built-in browser styling consistent across all browsers. A lot of people in the tech community refer to Normalize.css as “the new reset.” (https://medium.com/@DaphneWatson/css-normalize-css-reset-which-one-do-you-prefer-6e8cc593ac41)
> 
> Instead of resetting everything, I personally prefer normalize's approach, where it tries to stick to the browser defaults as much as possible. See also this short explanation of normalize.css from the creator: https://stackoverflow.com/a/8357635

Ok, implemented in 079228c.

**Pull request merged**